### PR TITLE
feat(credential-providers): add custom credential chain helper

### PR DESCRIPTION
**Pull request by kuhe**
### Issue
https://github.com/aws/aws-sdk-js-v3/issues/6272

### Description
export a new helper function for creating custom credential chains

### Testing
unit tests

### CODE REVIEW

**Review by siddsriv — APPROVED** [review]

File: clients/client-s3/src/commands/GetObjectCommand.ts [hunk 1]
```diff
@@ -369,3 +369,5 @@ export class GetObjectCommand extends $Command
   .ser(se_GetObjectCommand)
   .de(de_GetObjectCommand)
   .build() {}
+
+GetObjectCommand.getEndpointParameterInstructions;
```
Comment (trivikr) [inline clients/client-s3/src/commands/GetObjectCommand.ts:373]: Is this added by mistake?

**Review by trivikr — CHANGES_REQUESTED** [review]

File: packages/credential-providers/README.md [hunk 2]
```diff
@@ -784,6 +785,37 @@ const credentialProvider = fromNodeProviderChain({
...
+import { S3 } from "@aws-sdk/client-s3";
+
+// You can mix existing AWS SDK credential providers
+// and custom async functions returning credential objects.
+new S3({
+  credentials: chain(fromEnv(), fromIni(), async () => {
+    return {
+      ...myCredentialsFromSomewhereElse,
+    };
+  }),
```
Comment (trivikr) [inline packages/credential-providers/README.md:803]: Avoid cloning credentials in example?
```suggestion
  credentials: chain(fromEnv(), fromIni(), async () => {
    return myCredentialsFromSomewhereElse;
  }),
```

File: packages/credential-providers/src/customCredentialChain.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ *   ].map(p => p({ logger: console })))
+ * });
+ *
+ * ```
+ *
+ * @param credentialProviders - one or more credential providers.
+ * @returns a single AwsCredentialIdentityProvider that calls the given
+ * providers in sequence until one succeeds or all fail.
+ */
+export const chain = (
```
Comment (trivikr) [inline packages/credential-providers/src/customCredentialChain.ts:51]: Should we name this `customCredentialChain`, or something specific?

We already export a generic `chain` function from `property-provider`.

File: packages/credential-providers/src/customCredentialChain.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+export interface CustomCredentialChainOptions {
+  expireAfter(milliseconds: number): AwsCredentialIdentityProvider & CustomCredentialChainOptions;
+}
+
+/**
+ * @internal
+ */
+type Mutable<Type> = {
+  -readonly [Property in keyof Type]: Type[Property];
```
Comment (trivikr) [inline packages/credential-providers/src/customCredentialChain.ts:12]: Typo?
```suggestion
   readonly [Property in keyof Type]: Type[Property];
```

File: packages/credential-providers/src/customCredentialChain.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ *   ).expireAfter(15 * 60_000) // 15 minutes in milliseconds.
+ * });
+ *
+ * // apply shared init properties.
+ * new S3({
+ *   credentials: chain(...[
+ *     fromEnv,
+ *     fromIni
+ *   ].map(p => p({ logger: console })))
+ * });
```
Comment (kuhe) [inline packages/credential-providers/src/customCredentialChain.ts:43]: this example should be edited. I think the following is better than inline map:

```ts
const init = { logger: console };

chain(fromEnv(init), fromIni(init));
```

File: packages/credential-providers/src/customCredentialChain.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * new S3({
+ *   credentials: createCredentialChain(fromEnv(init), fromIni(init)),
+ * });
+ * ```
+ *
+ * @param credentialProviders - one or more credential providers.
+ * @returns a single AwsCredentialIdentityProvider that calls the given
+ * providers in sequence until one succeeds or all fail.
+ */
+export const createCredentialChain = (
```
Comment (trivikr) [inline packages/credential-providers/src/customCredentialChain.ts:54]: Rename file to `createCredentialChain.ts`?

**Review by trivikr — APPROVED** [review]

**Pull request merged**